modules/dag-l1/src/main/scala/org/tessellation/dag/l1/domain/transaction/ContextualTransactionValidator.scala

```diff
@@ -0,0 +1,282 @@
+
+          updatedBalance match {
+            case Left(_) =>
+              toProcess
+            case Right(_) if updatedTotalConsumedCap > 1 || memPool.isEmpty =>
+              toProcess
+            case Right(value) =>
+              loop(
+                memPool.tail,
+                value,
+                next.next(currentOrdinal),
+                currentOrdinal,
+                updatedTotalConsumedCap,
+                toProcess ++ memPool.headOption.toList
+              )
+          }
+        }
+
+        val allowedTransactions =
+          loop(memPool.appended(incomingTx), balance, lastMajorityTransactionOrdinal, currentOrdinal, 0.0d, List.empty)
+        allowedTransactions.find(_ === incomingTx)
+      }
+
+      private def getBalanceAffectedByTxs(
+        signedTx: Signed[Transaction],
+        txs: Option[SortedMap[TransactionOrdinal, StoredTransaction]],
+        latestBalance: Balance
+      ): F[Balance] =
+        txs match {
+          case Some(txs) =>
+            val aboveMajority = getTransactionsAboveMajority(txs)
+            signedTx.toHashed.map { hashedTx =>
+              val affectedTxs = resolveConflict(hashedTx, aboveMajority) match {
+                case NoConflict           => aboveMajority
+                case CanOverride          => aboveMajority.filter { case (ordinal, _) => ordinal < hashedTx.ordinal }
+                case CannotOverride(_, _) => aboveMajority
+              }
+              affectedTxs
+                .foldLeftM(latestBalance) {
+                  case (acc, curr) =>
+                    toAmount(curr.transaction.amount).plus(curr.transaction.fee).flatMap(acc.minus)
+                }
+                .getOrElse(Balance.empty)
+
+            }
+          case None => latestBalance.pure[F]
+        }
+
+      private def resolveConflict(
+        hashedTx: Hashed[Transaction],
+        txs: SortedMap[TransactionOrdinal, StoredTransaction]
+      ): ConflictResolveResult =
+        txs.get(hashedTx.ordinal) match {
+          case Some(WaitingTx(existing)) if existing.fee < hashedTx.fee => CanOverride
+          case Some(tx)                                                 => CannotOverride(tx.ref, TransactionReference.of(hashedTx))
+          case None                                                     => NoConflict
+        }
+
+      private def getSnapshotOrdinalOfMajorityTx(
+        txs: SortedMap[TransactionOrdinal, StoredTransaction]
+      ): SnapshotOrdinal =
+        txs.collectFirst { case (_, MajorityTx(_, snapshotOrdinal)) => snapshotOrdinal }.getOrElse(SnapshotOrdinal.MinValue)
+
+      private def getTransactionsAboveMajority(
+        txs: SortedMap[TransactionOrdinal, StoredTransaction]
+      ): SortedMap[TransactionOrdinal, StoredHashedTransaction] =
+        txs.collect {
+          case (ordinal, stored @ (_: WaitingTx | _: ProcessingTx | _: AcceptedTx)) =>
+            (ordinal, stored.asInstanceOf[StoredHashedTransaction])
```

Comment:
Is there a way to do this without using `asInstanceOf` ? If we are certain that this will never blow up then doesn't that mean the input type is going to be `StoredHashedTransaction` ?

Reply:
Is this an option?
```
case (ordinal, stored:StoredHashedTransaction @ (_: WaitingTx | _: ProcessingTx | _: AcceptedTx)) =>
  (ordinal, stored)
```

Reply:
> Is this an option?
> 
> ```
> case (ordinal, stored:StoredHashedTransaction @ (_: WaitingTx | _: ProcessingTx | _: AcceptedTx)) =>
>   (ordinal, stored)
> ```

It doesn't compile. Compilator is not smart enough to figure out that ann the `WaitingTx`, `ProcessingTx`, and `AcceptedTx` are all `StoredTransaction` but also `StoredHashedTransaction` (mixin for `StoredTransaction`)

Actually, that function should return all the transactions above the majority. After thinking twice I think that we can make an assumption that `MajorityTx` can exist only once in storage and it is always the `head` of storage. It is guaranteed by the logic which sets `MajorityTx` - it removes everything which is below `MajorityTx` and the new `MajorityTx` should have a higher ordinal than the previous one.

Then we can just take everything above literally:

```scala
txs.takeRight(txs.size -1)
```

or just

```scala
txs.filterNot { case (_, MajorityTx(_, _)) => true }
```

but it still is not smart enough to know that I narrowed the type by filtering out.

I think that the solution would be to:

```
txs.collect { case tx @ (_, _: StoredHashedTransaction) => tx }
```

Reply:
>that mean the input type is going to be StoredHashedTransaction?

I didn't want to make that assumption and I decided to be explicit because if at some point we decide to make `MajorityTx` to be `StoredHashedTransaction` as well then that logic will to be incorrect. Actually I think I will rename `StoredHashedTransaction` to just `NonMajorityTx` so name expresses the purpose/state instead of expressing the content.